# Análisis de las palabras de un texto.

En este ejercicio vamos ha realizar una serie de estadísticas simples sobre las palabras de un texto. En esta práctica solo se podrán usar las operaciones básica de cadenas de caracteres
`s[pos]`, función `len`, y concatenación de cadenas de caracteres (operador `+`) y de las listas las funciones de acceso y modificación `lst[pos]`, la longitud de una lista `len(lst)` y las operación `append`. 

Durante este ejercicio vamos a usar fragmentos del quijote. Ejecuta la siguiente casilla para leer el quijote en una constante.

In [1]:
with open('quijote.txt', encoding='UTF-8') as f:
    QUIJOTE = f.read()

In [2]:
pos = 26858 
print(QUIJOTE[pos:pos+1100])

En un lugar de la Mancha, de cuyo nombre no quiero acordarme, no ha mucho
tiempo que vivía un hidalgo de los de lanza en astillero, adarga antigua,
rocín flaco y galgo corredor. Una olla de algo más vaca que carnero,
salpicón las más noches, duelos y quebrantos los sábados, lantejas los
viernes, algún palomino de añadidura los domingos, consumían las tres
partes de su hacienda. El resto della concluían sayo de velarte, calzas de
velludo para las fiestas, con sus pantuflos de lo mesmo, y los días de
entresemana se honraba con su vellorí de lo más fino. Tenía en su casa una
ama que pasaba de los cuarenta, y una sobrina que no llegaba a los veinte,
y un mozo de campo y plaza, que así ensillaba el rocín como tomaba la
podadera. Frisaba la edad de nuestro hidalgo con los cincuenta años; era de
complexión recia, seco de carnes, enjuto de rostro, gran madrugador y amigo
de la caza. Quieren decir que tenía el sobrenombre de Quijada, o Quesada,
que en esto hay alguna diferencia en los autores q

In [3]:
pos = 26858 
QUIJOTE[pos:pos+1100]

'En un lugar de la Mancha, de cuyo nombre no quiero acordarme, no ha mucho\ntiempo que vivía un hidalgo de los de lanza en astillero, adarga antigua,\nrocín flaco y galgo corredor. Una olla de algo más vaca que carnero,\nsalpicón las más noches, duelos y quebrantos los sábados, lantejas los\nviernes, algún palomino de añadidura los domingos, consumían las tres\npartes de su hacienda. El resto della concluían sayo de velarte, calzas de\nvelludo para las fiestas, con sus pantuflos de lo mesmo, y los días de\nentresemana se honraba con su vellorí de lo más fino. Tenía en su casa una\nama que pasaba de los cuarenta, y una sobrina que no llegaba a los veinte,\ny un mozo de campo y plaza, que así ensillaba el rocín como tomaba la\npodadera. Frisaba la edad de nuestro hidalgo con los cincuenta años; era de\ncomplexión recia, seco de carnes, enjuto de rostro, gran madrugador y amigo\nde la caza. Quieren decir que tenía el sobrenombre de Quijada, o Quesada,\nque en esto hay alguna diferencia en

En primer lugar vamos a hacer una función que vamos a llamar `split` que va a dividir un texto en sus palabras. El texto nos lo van a proporcionar en una cadena de caracteres que se pasa como parámetro a la función. Una palabra va a ser cualquier secuencia de caracteres entre los signos de puntuación. Para obtener los signos de puntuación vamos a usar la constante `PUNCTUATION`. La función debe devolver una lista con las palabras que la componen. 

In [4]:
from testing import WrapperStr, WrapperLst, runtests

PUNCTUATION = WrapperStr(' ¡!"#$%&\'()*+,-./:;<=>¿?@[\\]^_`{|}~«»\n\t\r')

Para hacer esta función primero necesitamos una función auxiliar. La primera es `get_word` que admite como parámetro un texto (un `str`) y una posición. La función devuelve la siguiente palabra después de saltar los signos de puntuación y la posición siguiente a la palabra.

Ejemplos:
  'desta manera?: «Apenas había', 13 -----> 'Apenas', 22
  'desta manera?: «Apenas había', 23 -----> 'Había', 28
  'tomaba la\npodadera. Frisaba', 9  -----> 'podadera', 18

**Pista:** puedes hacer usar la función `is_in` que está en alguna hoja de ejercicios anterior.

In [5]:
### BEGIN SOLUTION
def is_in(s: str, ltr:str) -> bool:
    i = 0
    while i<len(s) and s[i]!=ltr:
        i = i + 1
    return i<len(s)



def get_word(txt:str, pos:int) -> str:
    i = pos
    while i<len(txt) and is_in(PUNCTUATION, txt[i]):
        i += 1
    word = ''
    while i<len(txt) and not is_in(PUNCTUATION, txt[i]):
        word += txt[i]
        i += 1
    return word, i
### END SOLUTION

In [6]:
def test_get_word():
    def wrapper(txt: str, pos: int):
        return get_word(WrapperStr(txt), pos)

    tests = [
        (('desta manera?: «Apenas había', 13), ('Apenas', 22)),
        (('desta manera?: «Apenas había', 23), ('había', 28)),
        (('tomaba la\npodadera. Frisaba', 9), ('podadera', 18)),
        (('tomaba la\npodadera. Frisaba', 27), ('', 27)),
        (('tomaba la\npodadera. Frisaba!.  ', 27), ('', 31))

    ]
    runtests(tests, wrapper)
test_get_word()

"('desta manera?: «Apenas había', 13)", "('Apenas', 22)"....OK.
"('desta manera?: «Apenas había', 23)", "('había', 28)"....OK.
"('tomaba la\npodadera. Frisaba', 9)", "('podadera', 18)"....OK.
"('tomaba la\npodadera. Frisaba', 27)", "('', 27)"....OK.
"('tomaba la\npodadera. Frisaba!.  ', 27)", "('', 31)"....OK.


Usando esa esa función, ya es fácil realizar la función `split` pedida. La función toma como parámetro una cadena de caracteres y devuelve una lista con las palabras que contiene. Fíjate en las últimas pruebas de la función anterior. No introduzcas la palabra vacía en la lista.

 

In [7]:
### BEGIN SOLUTION
def split(txt: str) -> list[str]:
    i = 0
    word_lst = []
    while i < len(txt):
        word, i = get_word(txt, i)
        if word!='':
            word_lst.append(word)
    return word_lst
### END SOLUTION

In [8]:
def test_split():
    def wrapper(txt: str):
        return sorted(split(WrapperStr(txt)))

    tests = [
        (('desta manera?: «Apenas había', ), ['Apenas', 'desta', 'había', 'manera']),
        (('tomaba la\npodadera. Frisaba', ), ['Frisaba', 'la', 'podadera', 'tomaba']),
        (('tomaba la\npodadera. ¡Frisaba!. ', ), ['Frisaba', 'la', 'podadera', 'tomaba']),
    ]
    runtests(tests, wrapper)
test_split()

"('desta manera?: «Apenas había',)", "['Apenas', 'desta', 'había', 'manera']"....OK.
"('tomaba la\npodadera. Frisaba',)", "['Frisaba', 'la', 'podadera', 'tomaba']"....OK.
"('tomaba la\npodadera. ¡Frisaba!. ',)", "['Frisaba', 'la', 'podadera', 'tomaba']"....OK.


Realiza una función que cambie una palabra a minúsculas. La palabra `CIGÜEÑA` la debe transformar a `cigüeña`.
Debes tener en cuenta las tildes, las eñes y las diéresis. 

In [9]:
### BEGIN SOLUTION
def lower(word: str) -> str:
    new = ''
    for ltr in word:
        if 'A' <= ltr <= 'Z':
            ltr = chr(ord('a') + ord(ltr) - ord('A'))
        elif ltr == 'Á':
            ltr = 'á'
        elif ltr == 'É':
            ltr = 'é'
        elif ltr == 'Í':
            ltr = 'í'
        elif ltr == 'Ó':
            ltr = 'ó'
        elif ltr == 'Ú':
            ltr = 'ú'
        elif ltr == 'Ü':
            ltr = 'ü'
        elif ltr == 'Ñ':
            ltr = 'ñ'
        new = new + ltr
    return new
### END SOLUTION

In [11]:
def test_lower():
    def wrapper(txt: str) -> str:
         return lower(WrapperStr(txt))

    tests = [
        (('CIGÜEÑA',), 'cigüeña'),
        (('Ángel',), 'ángel'),
        (('ÁÉÍÓÚÜÑ',), 'áéíóúüñ')
    ]
    runtests(tests, wrapper)
test_lower()

"('CIGÜEÑA',)", "cigüeña"....OK.
"('Ángel',)", "ángel"....OK.
"('ÁÉÍÓÚÜÑ',)", "áéíóúüñ"....OK.


Ahora debemos hacer una función llamada `different` que tenga como parámetro una lista de palabras. La función devuelve una lista con las palabras de la lista original sin repeticiones. Las palabras devueltas deben estar en minúsculas.
No se puede hacer una comprobación del tipo `if w not in lst`, para ello se puede usar la función `index` que vimos en clase

In [12]:
### BEGIN SOLUTION
def index(lst: list, elem) -> int:
    i = 0
    while i<len(lst) and lst[i]!=elem:
        i = i + 1
    if i<len(lst):
        pos = i
    else:
        pos = -1
    return pos

def different_words(words: list[str]) -> list[str]:
    different = []
    for w in words:
        lw = lower(w)
        if index(different, lw) == -1:
            different.append(lw)
    return different
### END SOLUTION

In [13]:
def test_different():
    def wrapper(words: list[str]) -> list[str]:
        newlst = list(map(WrapperStr, words))
        return sorted(different_words(newlst))

    tests = [
        (([],),[]),
        ((['Hola','Pepe', 'hola', 'pepe'],),['hola', 'pepe']),
        ((['Ángel','ángel', 'CIGÜEÑA', 'cigüeña'],),['cigüeña', 'ángel']),

    ]
    runtests(tests, wrapper)
test_different()

"([],)", "[]"....OK.
"(['Hola', 'Pepe', 'hola', 'pepe'],)", "['hola', 'pepe']"....OK.
"(['Ángel', 'ángel', 'CIGÜEÑA', 'cigüeña'],)", "['cigüeña', 'ángel']"....OK.


Vamos a realizar una función que calcule unas estadísticas simples: la media y la varianza. La función se llamará `mean_variance`, admite como parámetro una lista de números. Calculará ambos valores realizando una sola pasada sobre la lista, para ello podemos usar lo siguiente

$$
\begin{array}{l}
\displaystyle\frac{1}{n}\sum_{i=1}^{n} (x_i-\overline{x})^{2}=\\
\displaystyle\frac{1}{n}\sum_{i=1}^{n} x_i^{2}+\overline{x}^{2} - 2 x_{i}\overline{x}=\\
\displaystyle\left(\frac{1}{n}\sum_{i=1}^{n} x_i^{2}\right) +\
\frac{n}{n}\overline{x}^{2} - 2 \overline{x}^{2}\frac{1}{n}\sum_{i=1}^{n} x_i =\\
\displaystyle\left(\frac{1}{n}\sum_{i=1}^{n} x_i^{2}\right) +
\overline{x}^{2} - 2\overline{x}^{2}=\\
\displaystyle\left(\frac{1}{n}\sum_{i=1}^{n} x_i^{2}\right)- \overline{x}^{2}
\end{array}
$$


Donde $\overline{x}$ es la media de x.
Se hace un único bucle que calcula la suma de los valores y la suma de los cuadrados de los valores.
La función devuelve una tupla, el primer elemento es la media y el segundo la varianza.
Recuerda que es necesario que $n>0$.

In [14]:
### BEGIN SOLUTION
def mean_variance(serie: list[float]) -> (float, float):
    assert len(serie)>0, 'Error cannot calculate the mean an variance of an empty list'
    s = 0
    s_sq = 0
    for n in serie:
        s += n
        s_sq += n*n
    mean = s/len(serie)
    variance = s_sq/len(serie) - mean*mean
    return mean, variance
### END SOLUTION

# 

In [15]:
def test_mean_variance():
    def wrapper(serie: list[float]) -> (float, float):
        mean, var = mean_variance(serie)
        return round(mean, 6), round(var, 6)

    tests = [
        (([1.0, 2.0, 3.0],), (2.0, round(2/3, 6))),
        (([0.0, 1.0, 2.0, 3.0, 4.0],), (2.0, 2.0)),
    ]
    runtests(tests, wrapper)

    try:
        mean_variance([])
    except AssertionError:
        print('OK.')
    except:
        assert False, 'Se debe controlar que la lista no sea vacía'
test_mean_variance()

"([1.0, 2.0, 3.0],)", "(2.0, 0.666667)"....OK.
"([0.0, 1.0, 2.0, 3.0, 4.0],)", "(2.0, 2.0)"....OK.
OK.


La siguiente función que se debe llamar `get_lengths` tiene como parámetro una lista de palabras. Devuelve una lista con las longitudes de cada una de las palabras que aparecen en la lista original

**Ejemplo:** `['hola', 'mundo']` ----> `[4, 5]`

In [17]:
### BEGIN SOLUTION
def get_lengths(words: list[str]) -> list[int]:
    lengths = []
    for w in words:
        lengths.append(len(w))
    return lengths
### END SOLUTION

In [18]:
def test_get_lenghts():
    def wrapper(words: list[str]) -> list[str]:
        newlst = list(map(WrapperStr, words))
        return get_lengths(newlst)

    tests = [
        (([],),[]),
        ((['a', 'b', 'ab', 'abc', ''],), [1, 1, 2, 3, 0]),
        ((['cigüeña', 'ángel'],), [7, 5]),
    ]
    runtests(tests, wrapper)
test_get_lenghts()

"([],)", "[]"....OK.
"(['a', 'b', 'ab', 'abc', ''],)", "[1, 1, 2, 3, 0]"....OK.
"(['cigüeña', 'ángel'],)", "[7, 5]"....OK.


Finalmente podemos juntar todas las piezas. Debes hacer una función llamada `stats` que recibe como parámetro un texto. Debe devolver una lista de 6 elementos: `[all_len, all_mean, all_var, diff_len, all_mean, all_var]`, donde 
* `all_len` es el número total de palabras del texto.
* `all_mean` y `all_var` es la media y la varianza de las longitudes de todas las palabras del texto.
* `diff_len` es el número de palabras diferentes.
* `diff_mean` y `diff_var` es la media y la varianza de las longitudes de las palabras diferentes del texto.

**Ejemplo**: `'a ab abc A.AB,ABC'` ----> `[6, 2, 0.666667, 3, 2, 0.666667]`


In [19]:
### BEGIN SOLUTION
def stats(txt: str) -> (int, float, float, int, float, float):
    words = split(txt)
    assert len(words)>0, 'Error, el texto no debe ser vacío'
    wlen = get_lengths(words)
    wmean, wvar = mean_variance(wlen)
    diff = different_words(words)
    dlen = get_lengths(diff)
    dmean, dvar = mean_variance(dlen)
    return [len(words), wmean, wvar, len(diff), dmean, dvar]
### END SOLUTION

In [35]:
def test_stats():
    def wrapper(txt: str) -> (int, float, float, int, float, float):
        return list(map(lambda x: round(x,6), stats(WrapperStr(txt))))

    tests = [
        (('a',), [1,1,0,1,1,0]),
        (('a ab abc',), [3, 2, round(2/3,6), 3, 2, round(2/3,6)]),
        (('a ab abc A.AB,ABC',), [6, 2, round(2/3,6), 3, 2, round(2/3,6)]),
    ]
    runtests(tests, wrapper)
test_stats()

"('a',)", "[1, 1, 0, 1, 1, 0]"....OK.
"('a ab abc',)", "[3, 2, 0.666667, 3, 2, 0.666667]"....OK.
"('a ab abc A.AB,ABC',)", "[6, 2, 0.666667, 3, 2, 0.666667]"....OK.


Si has llegado hasta aquí podemos calcular las estadísticas del quijote. Pero con una advertencia, con lo que hemos visto hasta ahora tenemos un programa de complejidad cuadrática, no demasiado eficiente, y el el programa tarda bastante

In [50]:
from time import time
stime = time()
s = stats(QUIJOTE)
etime = time()
print(s)
print(f'La función ha tardado {etime-stime} segundos')

[384336, 4.308011739727738, 6.374145461916218, 23621, 7.795817281232801, 5.143737638007956]
La función ha tardado 132.9169008731842 segundos


El problema es que estamos usando listas para calcular las palabras diferentes. Si usamos una estructura más eficiente, que veremos más adelante, los tiempos se reducen drásticamente. Si usamos la siguiente función para calcular las palabras diferentes (que no se puede usar para el ejercicio previo), vemos como el tiempo se reduce drásticamente.

In [54]:
def different_words(words: list[str]) -> set[str]:
    different = set()
    for w in words:
        different.add(w)
    return list(different)

In [55]:
from time import time
stime = time()
s = stats(QUIJOTE)
etime = time()
print(s)
print(f'La función ha tardado {etime-stime} segundos')

[384336, 4.308011739727738, 6.374145461916218, 25269, 7.685820570659701, 5.291492718764793]
La función ha tardado 21.60258173942566 segundos
